config/config_test.go

```diff
@@ -62,17 +78,47 @@ func TestValidConfigs(t *testing.T) {
 func TestInvalidConfigs(t *testing.T) {
 	for k, v := range invalidConfigs {
 		t.Run(k, func(t *testing.T) {
-			filename := mkConfig(t, v)
-			_, err := ReadConfigFile(filename, "", "")
+			filename := mkConfig(t, t.TempDir(), v)
+			_, err := ReadConfigFile(filename, "", "", "")
 			if err == nil {
 				t.Fatal("this config must generate an error")
 			}
 		})
 	}
 }
 
-func mkConfig(t *testing.T, body string) string {
-	f, err := ioutil.TempFile(t.TempDir(), "config_test_*")
+func TestConfigDir(t *testing.T) {
+
+	dir := t.TempDir()
```

Comment:
This is more just about how I might approach this so don't feel like this needs to be done.

But for testing functionality like this where you're testing the same basic functionality but trying to capture different scenarios and edge cases it would be good to use table driven tests where you can easily add more scenarios.

Reply:
I would do something like define a struct that describes a single test:

```go
type configDirTest struct {
    files []string
    expectedSecrets []config.SecretType
    expectedAWS []config.AWSType
    expectedErr error
}
```

Then you can define your tests in a table:
```
var configDirTests = map[string]configDirTest{
    "test happy path!": {
        files: []string{
version: 3
secrets:
 - key: test
   path: /secret/test
   lifetime: static`,
version: 3
secrets:
 - key: test2
   path: /secret/test2
   lifetime: static
`
        },
        expectedSecrets: []config.SecretType{{Key: "test2"}, {Key: "test"}},
        expectedErr: nil,
    },
    // ...
}
```

And then finally you could use the assert library this project uses for other tests to check if the final config contains your expected objects: https://pkg.go.dev/github.com/stretchr/testify@v1.7.0/assert#Assertions.Contains

(See my comment about using the assert function regardless)

Reply:
Then you could also do stuff like test cases where the config reading fails using the expected error.

Reply:
oh I see, yeah that sounds great. I'll come back to it though to make some progress